ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,327 @@
+[comment]: <> (           * To avoid follow-up events, we need to either replay only or use a NoopWriter.)
+[comment]: <> (In the following sections we will go in more depth in some these mentioned parts of the chain.)
+
+### Compacting the log
+
+In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific point, such that we can compact our log. With building state on followers each node is responsible for taking their own snapshots and compacting their log.
+
+Part of the snapshot should be the last processed and last exported position for each exporter. The smallest number indicates until which position we can compact our log. In order to avoid running the exporters on all nodes and reduce the network load (since they export normally to an external system), the Leader has to periodically sync the last exported position to the followers. This is done via SWIM.
+
+The last processed position corresponds to the last processed command on the leader **or** the last applied event on the follower. 
+
+**TODO:** There is an issue with leaders compacting too early. Ask deepthi or nicolas about it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+[comment]: <> (       This is the technical portion of the ZEP. After reading it, a contributor should understand/know the following:)
+[comment]: <> (       - [ ] The impact of the changes on other features is clear.)
+[comment]: <> (       - [ ] The implementation is delineated)
+[comment]: <> (       - [ ] Known corner cases are listed and addressed)
+[comment]: <> (       The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.)
+[comment]: <> (       - [ ] Does the ZEP affect the official Zeebe distribution &#40;e.g. configuration, logging&#41;?)
+[comment]: <> (       - [ ] Does the ZEP require coordination with the platform team?)
+[comment]: <> (       - [ ] Does the ZEP require coordination with the Operate team?)
+
+**TODO:** After doing prototype rewrite this section.
+
+We can't reuse our current ReprocessingStateMachine, since it is written in a way such it doesn't support continuously replay. Currently, it is a one time thing. It scans the log for the last source position to determine, when it will stop with the replay. Furthermore, it collects error events to make sure the records are failing on reprocessing as well. Should be fixed with https://github.com/camunda-cloud/zeebe/issues/7270. We need a new ReplayStateMachine, which only consumes events continuously. The leader is the only one who is allowed to write commands and follow-up events, which ensures that the follower will not be faster than the leader.
+
+       
+## Compatibility
+
+We see no issues with this proposal.
```

Comment:
How to handle different versions between leader and follower? 
At some point, we want to support rolling updates. We should have an idea on how to handle these situations.

Reply:
Good Point. I probably need to add a section for that. But I'm also not sure whether we know how we handle it currently. :sweat_smile:

Reply:
Isn't that also a problem currently?

Reply:
@deepthidevaki I thought as well but was not sure :sweat_smile:

Reply:
Yes. It is already a kind of problem if there is an unlucky leader change during the update. 
But with this change, we will run definitely into this problem because all nodes process/replay the records.